plugins/woocommerce/src/StoreApi/Utilities/AgenticCheckoutUtils.php

```diff
@@ -0,0 +1,342 @@
+						'code'    => 'invalid_product_id',
+						'message' => __( 'Product ID must be numeric.', 'woocommerce' ),
+						'param'   => '$.items[' . $index . '].id',
+					],
+					400
+				);
+			}
+
+			$product_id = (int) $item['id'];
+			$quantity   = (int) $item['quantity'];
+
+			try {
+				$cart_controller->add_to_cart(
+					[
+						'id'       => $product_id,
+						'quantity' => $quantity,
+					]
+				);
+			} catch ( \Exception $e ) {
+				return self::create_error_response_from_exception( $e, $index );
+			}
+		}
+
+		return null;
+	}
+
+	/**
+	 * Create error response from exception.
+	 *
+	 * @param \Exception $exception The exception.
+	 * @param int        $item_index The item index in the request.
+	 * @return \WP_REST_Response Error response.
+	 */
+	public static function create_error_response_from_exception( \Exception $exception, $item_index ) {
+
+		$error_message = $exception->getMessage();
+		$error_code    = ErrorCode::INVALID; // Default fallback.
+
+		// Check if it's a RouteException with a specific error code.
+		if ( $exception instanceof RouteException ) {
+			$wc_error_code = $exception->getErrorCode();
+
+			// Map WooCommerce error codes to Agentic Commerce Protocol error codes.
+			switch ( $wc_error_code ) {
+				case 'woocommerce_rest_product_out_of_stock':
+				case 'woocommerce_rest_product_partially_out_of_stock':
+					$error_code = ErrorCode::OUT_OF_STOCK;
+					break;
+				case 'woocommerce_rest_product_not_purchasable':
+				case 'woocommerce_rest_cart_invalid_product':
+				case 'woocommerce_rest_invalid_product_id':
+					$error_code = ErrorCode::INVALID;
+					break;
+				default:
+					// Keep default 'invalid' for unknown error codes.
+					break;
+			}
+		}
+
+		return new \WP_REST_Response(
+			[
+				'type'    => ErrorType::INVALID_REQUEST,
+				'code'    => $error_code,
+				'message' => $error_message,
+				'param'   => '$.items[' . $item_index . ']',
+			],
+			400
+		);
+	}
+
+	/**
+	 * Set buyer data on customer.
+	 *
+	 * @param array        $buyer Buyer data.
+	 * @param \WC_Customer $customer Customer instance.
+	 */
+	public static function set_buyer_data( $buyer, $customer ) {
+		if ( isset( $buyer['first_name'] ) ) {
+			$first_name = wc_clean( wp_unslash( $buyer['first_name'] ) );
+			$customer->set_billing_first_name( $first_name );
+			$customer->set_shipping_first_name( $first_name );
+		}
+
+		if ( isset( $buyer['last_name'] ) ) {
+			$last_name = wc_clean( wp_unslash( $buyer['last_name'] ) );
+			$customer->set_billing_last_name( $last_name );
+			$customer->set_shipping_last_name( $last_name );
+		}
+
+		if ( isset( $buyer['email'] ) ) {
+			$email = sanitize_email( wp_unslash( $buyer['email'] ) );
+			if ( is_email( $email ) ) {
+				$customer->set_billing_email( $email );
+			}
+		}
+
+		if ( isset( $buyer['phone_number'] ) ) {
+			$phone = wc_clean( wp_unslash( $buyer['phone_number'] ) );
+			$customer->set_billing_phone( $phone );
```

Comment:
If you're conflating billing and shipping addresses, note that shipping has a phone number also.